staging/src/k8s.io/client-go/tools/cache/delta_fifo.go

```diff
@@ -310,13 +303,6 @@ func (f *DeltaFIFO) queueActionLocked(actionType DeltaType, obj interface{}) err
 		return KeyError{obj, err}
 	}
 
-	// If object is supposed to be deleted (last event is Deleted),
-	// then we should ignore Sync events, because it would result in
```

Comment:
We should ignore sync events *if the UID hasn't changed* and not ignore them *if it has changed*.

The bug this is talking about is when sync resurrected an object that was already deleted, and then it lived on forever as a zombie.

Reply:
Folks in other comments are saying that there's no race, but I would expect the queue getting drained depends on the consumer, but the sync event depends on the producer.

In theory, if we've observed the deletion event, we shouldn't be able to then see the object in a relist, but that depends on apiserver & etcd getting things exactly right. Which they should, nowadays. Maybe the best response to this is to detect it and panic, to make sure we get a bug report if it happens.

Reply:
Separate protections (and covering unit tests) were already added for the resync case. If a relist time travels back in time, it will replay history, but that is a separate bigger issue and applies to more than just the initial sync event queued up.

Reply:
As long as that condition results in a panic (we REALLY need to know if it happens) now, I'm good. I'm pretty sure this condition is why the code is like it is. Maybe it was always just a theoretical problem.

Reply:
The original code combined replace and resync flows. This protection was added in as an attempt to fix the resync issue, which introduced a bug in the relist flow, and didn't completely solve the resync issue (hence the follow up separation and resync-specific fix in https://github.com/kubernetes/kubernetes/pull/32125, which is still present/effective)

> As long as that condition results in a panic (we REALLY need to know if it happens) now, I'm good

I  don't think it's DeltaFIFOs job to do that. Accepting a list of items in Replace, setting them in the store, and not propagating a Sync event for one seems like a pretty contained bug to me.

Also, DeltaFIFO would only be able to recognize that scenario if the Deleted event hadn't been popped yet, so it's not like it would provide great protection from that condition. For every time travel caught that way, it would be more likely the Deleted event had already been popped and a new sync event would be happily delivered. There's a whole class of time-travel bugs that that wouldn't catch (sync followed by updates from the past with no delete involved)

Reply:
/hold
just to make sure we're on the same page. would like @smarterclayton's opinion as well

Reply:
I went through this code with fresh mind now and it LGTM.

I also agree with @liggitt comment about - there can be other time-travel bugs. So there is no point in catching some of them without catching others.

I think we should proceed with this PR.

Reply:
talked with daniel and clayton offline

/hold cancel